Framework/DataHandling/src/LoadPSIMuonBin.cpp

```diff
@@ -356,185 +472,304 @@ void LoadPSIMuonBin::assignOutputWorkspaceParticulars(
-  logAlg->setProperty("LogName", "Mon Deviation");
-  logAlg->setProperty("LogText", m_header.monDeviation);
-  logAlg->executeAsChildAlg();
+  addToSampleLog("Mon Deviation", m_header.monDeviation, outputWorkspace);
 
   if (m_header.realT0[0] != 0) {
-    for (const float &i : m_header.realT0) {
-      if (i == 0)
+    // 16 is the max size of realT0
+    for (auto i = 0u; i < 16; ++i) {
+      if (m_header.realT0[i] == 0)
         break;
-      logAlg->setProperty("LogType", "String");
-      logAlg->setProperty("LogName", "realT0 + i");
-      logAlg->setProperty("LogText", std::to_string(i));
+      addToSampleLog("realT0 " + std::to_string(i), m_header.realT0[i],
+                     outputWorkspace);
+    }
+  }
+
+  // Read in the temperature file if provided/found
+  try {
+    readInTemperatureFile(outputWorkspace);
+  } catch (std::invalid_argument &e) {
+    g_log.warning("Temperature file could not be loaded: " +
+                  std::string(e.what()));
+  } catch (std::runtime_error &e) {
+    g_log.warning("Temperature file could not be loaded:" +
+                  std::string(e.what()));
+  }
+}
+
+void LoadPSIMuonBin::processTitleHeaderLine(const std::string &line) {
+  bool titlesFound = false;
+  std::string foundTitles = "";
+  for (const auto &charecter : line) {
+    if (charecter == ':') {
+      titlesFound = true;
+      continue;
+    } else if (titlesFound) {
+      foundTitles += charecter;
+    }
+  }
+  boost::trim(foundTitles);
+  if (foundTitles.find("\\") == std::string::npos) {
+    boost::split(m_tempHeader.titles, foundTitles, boost::is_any_of(" "));
+    m_tempHeader.delimeterOfTitlesIsBackSlash = false;
+  } else {
+    boost::split(m_tempHeader.titles, foundTitles, boost::is_any_of("\\"));
+    m_tempHeader.delimeterOfTitlesIsBackSlash = true;
+  }
+} // namespace DataHandling
+
+void LoadPSIMuonBin::processDateHeaderLine(const std::string &line) {
+  // Assume the date is added in the same place as always
+  auto date = line.substr(2, 11);
+  auto _time = line.substr(14, 8);
+  m_tempHeader.startDateTime = getFormattedDateTime(date, _time);
+}
+
+void LoadPSIMuonBin::processHeaderLine(const std::string &line) {
+  if (line.find("Title") != std::string::npos) {
+    // Find sample log titles from the header
+    processTitleHeaderLine(line);
+
+  } else if (std::find(std::begin(psiMonths), std::end(psiMonths),
+                       line.substr(5, 3)) != std::end(psiMonths)) {
+    // If the line contains a Month in the PSI format then assume it conains a
+    // date on the line
+    processDateHeaderLine(line);
+  }
+}
+
+void LoadPSIMuonBin::readInTemperatureFileHeader(const std::string &contents) {
+  const int uselessLines = 6;
+  int lineNo = 0;
+  std::string line = "";
+  for (const auto charecter : contents) {
+    if (charecter == '\n') {
+      if (line[0] == '!' && lineNo > uselessLines) {
+        processHeaderLine(line);
+      } else if (line[0] != '!') {
+        break;
```

Comment:
is there a way to avoid using break?

Reply:
If the break isn't present it will iterate through the entire file and perform no task except growing a string every line and overwriting it with an empty string. I can't see a way around this at some point.